service_linux.go

```diff
@@ -182,6 +182,7 @@ func (sb *sandbox) addLBBackend(ip, vip net.IP, fwMark uint32, ingressPorts []*P
 		if sb.ingress {
 			filteredPorts = filterPortConfigs(ingressPorts, false)
 			if err := programIngress(gwIP, filteredPorts, false); err != nil {
+				filterPortConfigs(ingressPorts, true)
```

Comment:
I believe that potentially that is not enough, it will depend on where the failure happened... if that fails in the middle there can be some half configured stuff left behind

Reply:
I concur.   The programIngress() function does not fail gracefully or leave the system in a consistent state if it fails, unfortunately.  These changes are necessary for a successful rollback, but insufficient.

Reviewing programIngress() a bit I see there are several different stages of changes:

1. first ingress programming after startup

1. basic iptables setup required for per-port configuration

1. per-ingress port rule add/deletion consisting of:
    a. DNAT rule programming
    b. SNAT rule programming
    c. filter bypass for the given port
    d. setting up a listening proxy (for reasons I don't quite get yet)

Failures in 1. never abort processing:  they only flag the error.  So no problem there.  Failures in 2. do abort processing, but every time programIngress(...true) /* add */ is invoked, 2 checks each step to avoid clobbering existing work.  So if failure happens here, we are *probably* ok to just undo the filterPortConfigs().

Failures in part 3. are the problem in several ways.  We could have gotten through some, but not all of the ports.  In that case we should either undo the operation we just did on the port configurations that *were* successful, *or* return a list of ports that we were *not* successful in programming.   Those could then be passed to filterPortConfigs() to undo their tracking.  The latter solution seems simpler and cleaner to me.

Unfortunately, that's not quite sufficient either, because note that programming each port is a multi-stage process.  If the programming fails in the middle of one of the stages and we want to make things clean, we have to *undo* the parts that *did* succeed.  (e.g. if 3c fails, we need to undo parts 3a and 3b for that port to leave things in a consistent state.)

But even *that* isn't quite sufficient, because it depends on whether this was for add or delete and the *reason* for the failure.  If the operation was an "add", then removing the partial add makes sense.  But if the operation was a remove and it *failed* because for some reason the entry wasn't there, then adding it back would be an error.   Finally, there is the question of what to do if for some bizarre reason *rolling back* the partial port configuration fails.  Ugh.  :(

I think the goal of this patch is definitely worthwhile, but making such a solution robust is going to be tricky to say the least.